core/embed/rust/src/storage/cache.rs

```diff
@@ -0,0 +1,149 @@
+use core::ops::Deref;
+
+use spin::{Mutex, MutexGuard};
+use zeroize::Zeroizing;
+
+use crate::{lru::LruCache, trezorhal::random};
+
+pub enum CacheError {
+    InvalidSessionId,
+    InvalidValue,
+}
+
+pub struct Cache {
+    sessions: LruCache<SessionCache, 10>,
+    active_id: Option<SessionId>,
+}
+
+impl Cache {
+    pub const fn new() -> Self {
+        Self {
+            sessions: LruCache::new(),
+            active_id: None,
+        }
+    }
+
+    pub fn global() -> MutexGuard<'static, Self> {
+        static GLOBAL_CACHE: Mutex<Cache> = Mutex::new(Cache::new());
+
+        GLOBAL_CACHE.lock()
+    }
+
+    pub fn start_session(&mut self, received_session_id_bytes: Option<&[u8]>) -> SessionId {
+        // If we have received a session ID, take a look to the cache and return an ID
+        // of existing entry.
+        let received_session_id =
+            received_session_id_bytes.and_then(|bytes| SessionId::try_from(bytes).ok());
+        if let Some(received_id) = received_session_id {
+            if self.sessions.find(|s| s.id == Some(received_id)).is_some() {
+                return received_id;
+            }
+        }
+        // Either we haven't received an ID, or the entry doesn't exist (because of
+        // previous eviction). Create a new session.
+        let id = SessionId::random();
+        self.sessions.insert(SessionCache::new(id));
+        self.active_id = Some(id);
+        id
+    }
+
+    pub fn end_session(&mut self) {
+        if let Some(active_id) = self.active_id {
+            if let Some(active_session) = self.sessions.find(|s| s.id == Some(active_id)) {
+                active_session.clear();
+            }
+            self.active_id = None;
+        }
+    }
+
+    pub fn has_active_session(&self) -> bool {
+        self.active_id.is_some()
+    }
+}
+
+pub struct SessionCache {
+    id: Option<SessionId>,
+    fields: Option<SessionFields>,
+}
+
+impl SessionCache {
+    fn new(id: SessionId) -> Self {
+        Self {
+            id: Some(id),
+            fields: Some(SessionFields::default()),
+        }
+    }
+
+    fn clear(&mut self) {
```

Comment:
rather than clearing the cache entry itself, let's add a `drop(predicate)` to the cache?

Reply:
That would involve a rather large memcpy in this case, for no other benefit.